syntax = "proto3";

package payload.runtime.config;

import "google/protobuf/duration.proto";

/*
  Payload Manager Runtime Configuration

  This describes how a single node runs:
  storage tiers, database backend, worker behavior, etc.

  This file MUST NOT be used by clients.
*/

message ServerConfig {
  string bind_address = 1; // "0.0.0.0:50051"
}

// ------------------------------------------------------------------
// Database
// ------------------------------------------------------------------

message PostgresConfig {
  string connection_uri = 1;
  uint32 max_connections = 2;
}

message SqliteConfig {
  string path = 1;
  bool wal_mode = 2;
}

message DatabaseConfig {
  oneof backend {
    PostgresConfig postgres = 1;
    SqliteConfig sqlite = 2;
  }
}

// ------------------------------------------------------------------
// Storage tiers
// ------------------------------------------------------------------

message RamTierConfig {
  uint64 capacity_bytes = 1;
  bool use_hugepages = 2;
}

message DiskTierConfig {
  string root_path = 1;
  uint64 capacity_bytes = 2;
  bool fsync = 3;
}

message ObjectStoreConfig {
  string endpoint = 1;
  string bucket = 2;
  string access_key = 3;
  string secret_key = 4;
  bool tls = 5;
}

message GpuDeviceConfig {
  uint32 device_id = 1;
  uint64 capacity_bytes = 2;
}

message GpuTierConfig {
  repeated GpuDeviceConfig devices = 1;
}

message StorageConfig {
  RamTierConfig ram = 1;
  DiskTierConfig disk = 2;
  ObjectStoreConfig object = 3;
  GpuTierConfig gpu = 4;
}

// ------------------------------------------------------------------
// Workers
// ------------------------------------------------------------------

message SpillWorkerConfig {
  uint32 threads = 1;
  google.protobuf.Duration retry_backoff = 2;
}

message LeaseConfig {
  google.protobuf.Duration default_lease = 1;
  google.protobuf.Duration max_lease = 2;
}

message LoggingConfig {
  string level = 1;
  string pattern = 2;
  bool include_trace_context = 3;
}

// ------------------------------------------------------------
// Observability
// ------------------------------------------------------------

// OTLP transport selection
enum OtlpTransport {
  OTLP_TRANSPORT_UNSPECIFIED = 0; // Runtime default (gRPC)
  OTLP_TRANSPORT_GRPC = 1;        // OTLP over gRPC
  OTLP_TRANSPORT_HTTP = 2;        // OTLP over HTTP
}

message ObservabilityConfig {

  // Master enable switches
  bool metrics_enabled = 1;
  bool tracing_enabled = 2;
  bool logs_enabled = 3;

  // OTLP endpoint & transport selection
  string otlp_endpoint = 4;
  OtlpTransport transport = 5;

  // Tracing configuration
  message TracingConfig {

    // Tracing sampling intent
    enum TraceHint {
      TRACE_HINT_UNSPECIFIED = 0;
      TRACE_HINT_DEFAULT = 1;
      TRACE_HINT_ALWAYS = 2;
      TRACE_HINT_NEVER = 3;
    }

    TraceHint trace_hint = 1;

    // Trace processing strategy
    enum TraceProcessorType {
      TRACE_PROCESSOR_UNSPECIFIED = 0;
      TRACE_PROCESSOR_SIMPLE = 1;
      TRACE_PROCESSOR_BATCH = 2;
    }

    TraceProcessorType processor = 2;

    // Batch trace processor tuning
    message BatchConfig {
      uint32 max_queue_size = 1;
      uint32 max_export_batch_size = 2;
      uint32 schedule_delay_ms = 3;
      uint32 export_timeout_ms = 4;
    }

    BatchConfig batch = 3;

    // Span granularity controls
    bool stage_spans_enabled = 4;
    bool queue_spans_enabled = 5;
    bool record_spans_enabled = 6;

    // Trace attribute verbosity
    enum AttributeLevel {
      ATTRIBUTES_UNSPECIFIED = 0;
      ATTRIBUTES_MINIMAL = 1;
      ATTRIBUTES_STANDARD = 2;
      ATTRIBUTES_VERBOSE = 3;
    }

    AttributeLevel attribute_level = 7;

    uint32 latency_budget_ms = 8;
  }

  TracingConfig tracing = 6;

  // Metrics configuration
  message MetricsConfig {

    // Instrument controls for metrics emitted by this runtime.
    bool request_metrics_enabled = 1;
    bool spill_metrics_enabled = 2;
    bool tier_occupancy_metrics_enabled = 3;

    // Cost / cardinality controls
    bool request_latency_histograms_enabled = 4;
    bool route_labels_enabled = 5;
    bool tier_labels_enabled = 6;

    // Collection / export timing hints
    uint32 min_collection_interval_ms = 7;
    uint32 collection_interval_ms = 8;
    uint32 export_timeout_ms = 9;
  }

  MetricsConfig metrics = 7;

  // Logging configuration
  message LoggingConfig {

    enum LogProcessorType {
      LOG_PROCESSOR_UNSPECIFIED = 0;
      LOG_PROCESSOR_SIMPLE = 1;
      LOG_PROCESSOR_BATCH = 2;
    }

    LogProcessorType processor = 1;

    message BatchConfig {
      uint32 max_queue_size = 1;
      uint32 max_export_batch_size = 2;
      uint32 schedule_delay_ms = 3;
      uint32 export_timeout_ms = 4;
    }

    BatchConfig batch = 2;
  }

  LoggingConfig logging = 8;
}

// ------------------------------------------------------------------
// Top-level
// ------------------------------------------------------------------

message RuntimeConfig {
  ServerConfig server = 1;
  DatabaseConfig database = 2;
  StorageConfig storage = 3;
  SpillWorkerConfig spill_workers = 4;
  LeaseConfig leases = 5;
  LoggingConfig logging = 6;
  ObservabilityConfig observability = 7;
}
